Towards #3221 - serialized task updates

meichstedt opened the pull request.

kolloch commented on src/main/scala/mesosphere/marathon/core/launcher/impl/OfferProcessorImpl.scala:
```diff
@@ -103,9 +104,9 @@ private[launcher] class OfferProcessorImpl(
         nextOp.oldTask match {
           case Some(existingTask) =>
-            taskCreationHandler.created(existingTask).map(_ => ())
+            taskCreationHandler.created(TaskStateOp.Create(existingTask)).map(_ => ())
```
I would still prefer an explicit `TaskStateOp.Revert` operation.

meichstedt replied:
changed.

kolloch commented on src/main/scala/mesosphere/marathon/core/launcher/impl/OfferProcessorImpl.scala:
```diff
@@ -121,20 +122,10 @@ private[launcher] class OfferProcessorImpl(
-      }
-
-      persistedOp.map(_ => Some(taskOpWithSource))
+      log.info("Save task [{}]", taskOpWithSource.taskId)
```
you can already change the logging "Persisting TaskStateOp"

In fact, you can already rename the `newTask` field -- nobody is working on this concurrently. At least it confused me...

Have you documented that you want to change taskCreationHandler.created somewhere? Do you keep track of planned renames somewhere? We have this issue for that: https://github.com/mesosphere/marathon/issues/3311

meichstedt replied:
now renamed – there was a fixme which I removed, I've added it to #3311.

kolloch commented on src/main/scala/mesosphere/marathon/core/launcher/impl/TaskLauncherImpl.scala:
```diff
@@ -37,8 +37,8 @@ private[launcher] class TaskLauncherImpl(
       val launchCount = taskOps.count {
-        case TaskOp.Launch(_, _, _, _) => true
-        case _                         => false
+        case _: TaskOp.Launch => true
```
better :+1:

kolloch commented on src/main/scala/mesosphere/marathon/core/launcher/impl/TaskOpFactoryHelper.scala:
```diff
@@ -13,21 +13,31 @@ class TaskOpFactoryHelper(
+
+  def launchOnReservation(
+    taskInfo: Mesos.TaskInfo,
+    newTask: TaskStateOp,
```
could the type be more specific? LaunchOnReservation?

meichstedt replied:
Oh, yes. I also changed the oldTask type and the related type in `VolumeMatcher`

kolloch commented on src/main/scala/mesosphere/marathon/core/launcher/impl/TaskOpFactoryHelper.scala:
```diff
@@ -36,6 +46,6 @@ class TaskOpFactoryHelper(
       offerOperationFactory.createVolumes(frameworkId, newTask.taskId, localVolumes))
 
-    TaskOp.ReserveAndCreateVolumes(newTask, resources, localVolumes, oldTask, createOperations)
+    TaskOp.ReserveAndCreateVolumes(newTask, resources, localVolumes, createOperations)
```
jepp, you should be right, no old task should exist for this. Good to make that explicit :+1:

kolloch commented on src/main/scala/mesosphere/marathon/core/launcher/impl/TaskOpFactoryHelper.scala:
```diff
@@ -13,21 +13,31 @@ class TaskOpFactoryHelper(
   private[this] val offerOperationFactory = new OfferOperationFactory(principalOpt, roleOpt)
 
-  def launch(
+  def launchEphemeral(
```
I like the new more explicit "launch*" names, thanks! :+1:

kolloch commented on src/main/scala/mesosphere/marathon/core/task/bus/TaskStatusObservables.scala:
```diff
@@ -9,13 +9,17 @@ object TaskStatusObservables {
   case class TaskStatusUpdate(timestamp: Timestamp, taskId: Task.Id, status: MarathonTaskStatus) {
```
You can remove the `TaskStatusUpdate` class.

meichstedt replied:
Yay, removed. Just kept that in there while migrating.

kolloch commented on src/main/scala/mesosphere/marathon/core/task/bus/TaskStatusObservables.scala:
```diff
@@ -9,13 +9,17 @@ object TaskStatusObservables {
 trait TaskStatusObservables {
-  def forAll: Observable[TaskStatusUpdate]
-  def forAppId(appId: PathId): Observable[TaskStatusUpdate]
+  def forAll: Observable[TaskUpdate]
```
With this change you actually already moved pretty close to remove these TaskStatusObervables alltogether. Nice!

meichstedt replied:
Ha, I wasn't even aware of that :grin:

kolloch commented on src/main/scala/mesosphere/marathon/core/launchqueue/impl/AppTaskLauncherActor.scala:
```diff
@@ -221,41 +221,31 @@ private class AppTaskLauncherActor(
-        }
+  private[this] def receiveTaskUpdate: Receive = {
+    case TaskUpdate(stateOp, stateChange) =>
+      log.info("received TaskUpdate")
```
I think that I would get rid of this info log statement and promote the more specific log statements below to info if you want more visibility.

meichstedt replied:
Totally, I like the idea. Changed it!

kolloch commented on src/main/scala/mesosphere/marathon/core/launchqueue/impl/AppTaskLauncherActor.scala:
```diff
@@ -341,19 +331,23 @@ private class AppTaskLauncherActor(
           scheduleTaskOpTimeout(taskOp)
+
         case None =>
-          removeTask(taskId)
```
hmmm. "removeTask" is not really needed -- it was mostly defensive and maybe misleading. It was different in some historical state, I think. But then, we can actually just not do anything because the task should not be in the tasksMap. This branch is unnecessary, isn't it?

meichstedt replied:
Right, removed it.

kolloch commented on src/main/scala/mesosphere/marathon/core/leadership/LeadershipModule.scala:
```diff
@@ -36,11 +37,13 @@ object LeadershipModule {
   override def startWhenLeader(props: Props, name: String): ActorRef = {
-    require(!started, "already started")
+    log.info("startWhenLeader: {}", name + props.toString)
+    require(!started, s"already started: $name ${props.toString}")
```
the `toString` is implicit...

meichstedt replied:
thanks for finding this debugging-leakage :+1:

kolloch commented on src/main/scala/mesosphere/marathon/core/leadership/LeadershipModule.scala:
```diff
@@ -51,6 +54,9 @@ private[leadership] class LeadershipModuleImpl(
   private[this] lazy val coordinator_ = {
     require(!started, "already started")
+    log.info("LeadershipModule is now started")
+    //    throw new scala.RuntimeException("Sorry!")
```
Apology accepted. Now you can just delete that line.

meichstedt replied:
Ha ha ^_^

kolloch commented on src/main/scala/mesosphere/marathon/core/matcher/reconcile/impl/OfferMatcherReconciler.scala:
```diff
@@ -62,9 +62,14 @@ private[reconcile] class OfferMatcherReconciler(taskTracker: TaskTracker, groupR
+                  resources = spuriousResources.to[Seq]
                 )
               TaskOpWithSource(source(offer.getId), unreserveAndDestroy)
+            //          case (taskId, spuriousResources) if shouldRefreshLastSeen(taskId) =>
```
this can go, I guess

kolloch commented on src/main/scala/mesosphere/marathon/core/task/Task.scala:
```diff
@@ -83,39 +83,46 @@ object Task {
             startedAt = Some(now),
             mesosStatus = mesosStatus))
-        TaskStateChange.Update(updated)
+        TaskStateChange.Update(updated, Some(this))
```
I think that named parameters would make this a lot clearer.

meichstedt replied:
ack – renamed to `newState` and `oldState`, respectively

kolloch commented on src/main/scala/mesosphere/marathon/core/task/Task.scala:
```diff
@@ -83,39 +83,46 @@ object Task {
-
-        healthOrStateChange.getOrElse{
+      case TaskStateOp.MesosUpdate(_, taskStatus, now) =>
+        updatedHealthOrState(status.mesosStatus, taskStatus.mesosStatus).map { newStatus =>
```
Style (ignore me if you want)

if you handle both cases of an option, I prefer a `match`.

meichstedt replied:
I agree it is more readable

kolloch commented on src/main/scala/mesosphere/marathon/core/task/Task.scala:
```diff
@@ -83,39 +83,46 @@ object Task {
-        TaskStateChange.Failure("Unable to handle Timeout op on LaunchedEphemeral")
+      // FIXME (3221): If the task needs to recreated after a not-accepted taskOp, Create is used.
+      // that's neither nice nor obvious, so this should be improved
+      case TaskStateOp.Create(task) =>
```
I would like this to be renamed to "Revert"

kolloch replied:
And it could take an Option of a task such that it translates to an `Update` or an `Expunge`

meichstedt replied:
In order to keep the changes small I chose to go for
- `TaskStateOp.Revert(task: Task)` reverts to the given state
- `TaskStateOp.ForceExpunge(taskId: Task.Id)` expunges the given taskId

These 2 can be consolidated later on: at least no there are these 2 that explicitly handle recovery – I hope you're ok with that for now.

kolloch commented on src/main/scala/mesosphere/marathon/core/task/Task.scala:
```diff
@@ -169,44 +188,55 @@ object Task {
             startedAt = Some(now),
             mesosStatus = mesosStatus))
-        TaskStateChange.Update(updated)
+        TaskStateChange.Update(updated, Some(this))
```
named parameters?

meichstedt replied:
everywhere!

kolloch commented on src/main/scala/mesosphere/marathon/core/task/tracker/impl/TaskCreationHandlerAndUpdaterDelegate.scala:
```diff
@@ -3,50 +3,54 @@ package mesosphere.marathon.core.task.tracker.impl
 
 /**
-  * Implements the [[TaskCreationHandler]]/[[TaskUpdater]] traits by sending messages to the TaskTracker actor.
+  * Implements the [[TaskStateOpProcessor]] trait by sending messages to the TaskTracker actor.
```
`TaskOpProcessor` and `TaskStateOpProcessor` are very confusingly named since they do very similar things.

Is there a cleanup/renaming issue that already tracks this?

meichstedt replied:
Agree. I've added a bullet point to #3311

kolloch commented on src/main/scala/mesosphere/marathon/core/task/tracker/impl/TaskCreationHandlerAndUpdaterDelegate.scala:
```diff
@@ -3,50 +3,54 @@ package mesosphere.marathon.core.task.tracker.impl
   }
-  override def terminated(taskId: Task.Id): Future[_] = {
-    taskUpdate(taskId, TaskOpProcessor.Action.Expunge)
+  override def terminated(stateOp: TaskStateOp.ForceExpunge): Future[_] = {
```
Covered by cleanup issue to remove the specialized methods?

meichstedt replied:
yes

kolloch commented on src/main/scala/mesosphere/marathon/core/task/tracker/impl/TaskOpProcessorImpl.scala:
```diff
@@ -18,43 +19,55 @@ import scala.util.control.NonFatal
+      // FIXME (3221): Create and Expunge are mostly to restore states.
+      op match {
+        case op: TaskStateOp.Create              => updateIfNotExists(op.taskId, op.task)
+        case op: TaskStateOp.ForceExpunge        => updateExistingTask(op)
```
I am not sure if it is much better (especially with our code formatter) but you do now that you can group clauses, right? You basically only distinguish between updating and creating here.

meichstedt replied:
At least IDEA by default put annoying (yet sometimes helpful) warnings in there when I did that in the past, therefore I moved away from doing it completely.

kolloch commented on src/main/scala/mesosphere/marathon/core/task/tracker/impl/TaskOpProcessorImpl.scala:
```diff
@@ -18,43 +19,55 @@ import scala.util.control.NonFatal
+
 /**
   * Processes durable operations on tasks by
   *
```
Extend comment to cover new functionality.

meichstedt replied:
done.

kolloch commented on src/main/scala/mesosphere/marathon/core/task/tracker/impl/TaskOpProcessorImpl.scala:
```diff
@@ -64,51 +77,96 @@ private[tracker] object TaskOpProcessorImpl {
+    (stateOp, stateChange) match {
+      case (TaskStateOp.MesosUpdate(taskId, MarathonTaskStatus.WithMesosStatus(mesosStatus), now),
+        update: TaskStateChange.Update) =>
+        log.info("StateChange.Update -> notifying taskStatusUpdateSteps")
```
Unify this block with the next by also logging the class name of the taskStateOp?

meichstedt replied:
done.

kolloch commented on src/main/scala/mesosphere/marathon/core/task/tracker/impl/TaskOpProcessorImpl.scala:
```diff
@@ -64,51 +77,96 @@ private[tracker] object TaskOpProcessorImpl {
+  }
+
+  private[this] def notifyOthers(stateOp: TaskStateOp, stateChange: TaskStateChange): Future[Unit] = {
+    // compatibility: if it's a statusUpdate that resulted in a Update or Expunge, do as before
```
Am I missing something or are you only propagating mesos status updates?

We also need to propagate other updates -- at least to the launch queue (which will forward them to the `AppTaskLauncherActor`).

kolloch replied:
That should then also be somehow covered in unit/integration tests.

meichstedt replied:
Seems like I accidentally reverted that – this was the main point in all this refactoring …

I agree that this should be tested, I will add some tests.

kolloch commented on src/main/scala/mesosphere/marathon/core/task/tracker/impl/TaskOpProcessorImpl.scala:
```diff
@@ -129,20 +185,20 @@ private[tracker] class TaskOpProcessorImpl(
-      log.warn(
-        s"${op.taskId} of app [${op.taskId.appId}]: try to recover from failed ${op.action.toString}", cause
-      )
+      log.warn(s"${op.taskId} of app [${op.taskId.appId}]: try to recover from failed ${op.stateOp.toString}", cause)
```
`toString` is implicit and it is customary to not write it explicitly.

meichstedt replied:
removed

meichstedt commented:
Thanks @kolloch for your in-depth comments! I hope I've addressed them all. As noted above I have added tasks to #3311 – I will try to resolve those asap. I'd like to pull at least the TaskOp/TaskStateOp renaming into this PR, maybe we find some nice names tomorrow.

meichstedt commented:
Oh, in case you have more comments I will of course address them as well. Just realized the PR was still assigned to you :grinning:

kolloch commented on src/main/scala/mesosphere/marathon/core/task/tracker/impl/TaskOpProcessorImpl.scala:
```diff
@@ -18,97 +19,149 @@ import scala.util.control.NonFatal
-        }
+          taskTrackerRef ! TaskTrackerActor.TaskRemoved(op.taskId, TaskTrackerActor.Ack(op.sender, change))
+        }.flatMap { _ =>
+          notifyOthers(op.stateOp, change)
```
It is not (strictly) guaranteed that the message to the `TaskTracker` is processed before the steps for this change are processed. It would be nice, to guarantee that. The `TaskTrackerActor.Ack` construct was introduced to do exactly that for the other changes -- the `TaskTracker` would notify the requestor of this change of the completion AFTER it had applied the change to its in-memory task list.

kolloch replied:
As discussed, if someone outside of this package wants to update a task, the related future should only finish after the task tracker was updated and all steps executed sequentially for this update.

The steps themselves should only be executed after the TaskTracker has processed the related change. Therefore, it probably makes sense to call the step logic from the TaskTracker after it gets a change.

meichstedt replied:
Thanks for the suggestion, totally makes sense to put this one level deeper. I've bundled step processing into an own class and adjusted tha `TaskOpProcessorImpl`.

kolloch commented on src/main/scala/mesosphere/marathon/core/task/tracker/impl/TaskOpProcessorImpl.scala:
```diff
@@ -18,97 +19,149 @@ import scala.util.control.NonFatal
+    // compatibility: if it's a statusUpdate that resulted in a Update or Expunge, do as before
+
+    stateChange match {
+      case _: TaskStateChange.Update | _: TaskStateChange.Expunge =>
```
This logic is redundant. You only call the `notifyOthers` method for these cases so you don't really check for the relevant ops again.

meichstedt replied:
fixed.

kolloch commented on src/main/scala/mesosphere/marathon/core/task/bus/TaskStatusObservables.scala:
```diff
@@ -1,13 +1,14 @@
 object TaskStatusObservables {
-  case class TaskStatusUpdate(timestamp: Timestamp, taskId: Task.Id, status: MarathonTaskStatus) {
-    def appId: PathId = taskId.appId
+  case class TaskUpdate(stateOp: TaskStateOp, stateChange: TaskStateChange) {
```
I would prefer `TaskChanged`. 

* `TaskUpdate` is very close to `TaskStateOp.Update` and not restricted to that.
* The past tense suggests that it is a notification of something that happened already.

meichstedt replied:
Agree, it's easer to understand. My original renaming just removed the `Status` in between. To keep the naming consistent, I've also renamed vals containing a `TaskUpdate` to `taskUpdate`

kolloch commented on src/main/scala/mesosphere/marathon/core/task/bus/TaskStatusObservables.scala:
```diff
@@ -1,13 +1,14 @@
+import mesosphere.marathon.state.PathId
 import rx.lang.scala.Observable
 
 object TaskStatusObservables {
```
Suggestion: => TaskChangeObservables

meichstedt replied:
Makes sense. Also: `InternalTaskChangeEventStream`

kolloch commented on src/main/scala/mesosphere/marathon/core/task/Task.scala:
```diff
@@ -83,39 +83,48 @@ object Task {
-      // failure case: Timeout
-      case TaskStateOp.ReservationTimeout =>
-        TaskStateChange.Failure("Unable to handle Timeout op on LaunchedEphemeral")
+      case TaskStateOp.LaunchEphemeral(task) =>
```
Should not be called.

I wonder if we could make it more clear that LaunchEphemeral/Reserve/Revert should never be called on existing tasks.

Maybe subdivide TaskStateOp into two sub sealed traits: NewTaskStateOp, ChangeTaskStateOp. Then implement Task.update only for ChangeTaskStateOp?

meichstedt replied:
I'm absolutely with you. It bothered me several times that these case classes are handled differently, so yes, nice idea. Yet I'd like to defer this to a later clean up step as I fear that there are a lot of tedious test suite adjustments needed for this. I have added a sub task to #3311 for this.

kolloch commented on src/main/scala/mesosphere/marathon/core/task/Task.scala:
```diff
@@ -140,16 +149,30 @@ object Task {
+      // failure case
       case _: TaskStateOp.MesosUpdate =>
-        TaskStateChange.Failure("Unable to handle MesosUpdate op on Reserved task")
+        TaskStateChange.Failure("MesosUpdate on Reserved is not allowed")
```
In this case I would prefer the wording "unexpected" instead of "not allowed"

meichstedt replied:
done.

kolloch commented on src/main/scala/mesosphere/marathon/core/task/TaskStateOp.scala:
```diff
@@ -1,22 +1,59 @@
-  case class MesosUpdate(status: MarathonTaskStatus, now: Timestamp) extends TaskStateOp
-  case object ReservationTimeout extends TaskStateOp
+  /** Launch (aka create) an ephemeral task*/
+  // FIXME (3221): The type should be LaunchedEphemeral but that needs a lot of test adjustments
```
Is this tracked in a cleanup issue?

meichstedt replied:
Yes, see #3311

kolloch commented on src/main/scala/mesosphere/marathon/core/task/tracker/TaskCreationHandler.scala:
```diff
@@ -14,12 +14,12 @@ trait TaskCreationHandler {
     * that you generate unique IDs.
     */
-  def created(task: Task): Future[Task]
+  def created(taskStateOp: TaskStateOp): Future[Unit]
```
Is there a cleanup issue for consolidating this usages?

meichstedt replied:
#3311

kolloch commented on src/main/scala/mesosphere/marathon/core/task/update/impl/steps/ScaleAppUpdateStepImpl.scala:
```diff
@@ -3,37 +3,43 @@ package mesosphere.marathon.core.task.update.impl.steps
+  override def processUpdate(update: TaskUpdate): Future[_] = {
+
+    val terminalOrExpungedTask: Option[Task] = {
+      (update.stateOp, update.stateChange) match {
```
The `TaskUpdate` does not contain the oldTask option. Otherwise I would say that we need to scale if the oldTask contained a launch, and the new task option does not.

You should have covered all current cases, though.

meichstedt replied:
Yes, the oldTask is _hidden_ inside of `TaskChanged.stateChange` and is set if the stateChange is an Expunge. The match is needed because there's no Expunge in case it's a resident task. So basically we have either
- some StateOp that led to an Expunge OR
- a MesosStatusUpdate that led to either an Expunge or an Update(from LaunchedOnReservation to Reserved)

kolloch commented on src/main/scala/mesosphere/marathon/health/HealthCheckActor.scala:
```diff
@@ -182,6 +181,21 @@ class HealthCheckActor(
+    marathonSchedulerDriverHolder: MarathonSchedulerDriverHolder,
+    healthCheck: HealthCheck,
+    taskTracker: TaskTracker,
+    eventBus: EventStream): Props = {
```
:+1: for a props method

kolloch commented on src/main/scala/mesosphere/marathon/health/MarathonHealthCheckManager.scala:
```diff
@@ -63,9 +66,7 @@ class MarathonHealthCheckManager @Inject() (
-              Props(
-                classOf[HealthCheckActor],
-                app, driverHolder, scheduler, healthCheck, taskTracker, eventBus))
+              HealthCheckActor.props(app, driverHolder, healthCheck, taskTracker, eventBus))
```
thanks for making this type safe :+1: (or lightbend??)

kolloch commented on src/test/scala/mesosphere/marathon/LaunchQueueModuleTest.scala:
```diff
@@ -205,7 +205,7 @@ class LaunchQueueModuleTest
   after {
     verifyNoMoreInteractions(appRepository)
-    verifyNoMoreInteractions(taskTracker)
+    //    verifyNoMoreInteractions(taskTracker)
```
uncomment again?

meichstedt replied:
done.

kolloch commented on src/test/scala/mesosphere/marathon/core/task/tracker/impl/TaskOpProcessorImplTest.scala:
```diff
@@ -22,45 +37,6 @@ class TaskOpProcessorImplTest
   // ignored by the TaskOpProcessorImpl
   val deadline = Timestamp.zero
 
-  test("process noop") {
```
why did you delete these tests?

meichstedt replied:
There is no `TaskStateOp.Noop` or adequate. The case when NoChange is needed is handled in _process statusUpdate with NoChange_

kolloch commented on src/test/scala/mesosphere/marathon/core/task/tracker/impl/StatusUpdateActionResolverTest.scala:
```diff
@@ -1,57 +0,0 @@
-    extends FunSuite with Mockito with GivenWhenThen with ScalaFutures with Matchers {
-  import scala.concurrent.ExecutionContext.Implicits.global
-
-  test("an update for a non-existing tasks is mapped to fail") {
```
why did you delete this test or is it covered elsewhere?

meichstedt replied:
That's now located in `TaskStateOpResolverTest`

kolloch commented on src/test/scala/mesosphere/marathon/core/task/tracker/impl/TaskStateOpResolverTest.scala:
```diff
@@ -23,15 +25,129 @@ class TaskStateOpResolverTest
-    val stateChange = f.stateOpResolver.resolve(TaskStateOp.ForceExpunge(taskId)).futureValue
+    val stateChange = f.stateOpResolver.resolve(TaskStateOp.ForceExpunge(f.notExistingTaskId)).futureValue
+
+    Then("getTaskAsync is called")
```
consistent with the rest of the tests -- but the method is not called getTaskAsync anymore.

meichstedt replied:
right, fixed

kolloch commented on src/test/scala/mesosphere/marathon/core/task/tracker/impl/TaskStateOpResolverTest.scala:
```diff
@@ -23,15 +25,129 @@ class TaskStateOpResolverTest
+    f.verifyNoMoreInteractions()
+  }
+
+  test("ForceExpunge fails if task does not exist") {
```
What is the difference to the last test?

Also, if we use ForceExpunge as a revert and we detect that the task does not exist. Is that really an error? (Corner-case, ignore if you want for now)

meichstedt replied:
There's no difference, so I've removed it. As for the Force-Expunge, you're right. I think a ForceExpunge on a non-existent task should result in a NoChange

kolloch commented on src/test/scala/mesosphere/marathon/core/task/tracker/impl/TaskStateOpResolverTest.scala:
```diff
@@ -23,15 +25,129 @@ class TaskStateOpResolverTest
+    Given("a non existing taskId")
+    f.taskTracker.task(f.notExistingTaskId) returns Future.successful(None)
+
+    When("A ForceExpunge is scheduled with that taskId")
```
Text incorrect

meichstedt replied:
fixed.

kolloch commented on src/test/scala/mesosphere/marathon/core/task/tracker/impl/TaskStateOpResolverTest.scala:
```diff
@@ -23,15 +25,129 @@ class TaskStateOpResolverTest
+    Given("an existing task")
+    f.taskTracker.task(f.existingReservedTask.taskId) returns Future.successful(Some(f.existingReservedTask))
+
+    When("A create is scheduled with that taskId")
```
incorrect text

meichstedt replied:
fixed

kolloch commented on src/main/scala/mesosphere/marathon/core/launcher/impl/TaskOpFactoryHelper.scala:
```diff
@@ -21,14 +21,14 @@ class TaskOpFactoryHelper(
     taskInfo: Mesos.TaskInfo,
-    newTask: TaskStateOp,
-    oldTask: Task): TaskOp.Launch = {
+    newTask: TaskStateOp.LaunchOnReservation,
```
the types read very well, I like it! The parameter newTask should be renamed to taskOp.

kolloch commented on src/test/scala/mesosphere/marathon/core/task/tracker/impl/TaskOpProcessorImplTest.scala:
```diff
@@ -411,6 +413,114 @@ class TaskOpProcessorImplTest
+  test("LaunchEphemeral -> Update results processing update steps") {
+    val f = new Fixture
+    verifyStepsAreProcessed(
+      f = new Fixture,
```
`f = f`? or why are you creating a new fixture here?

kolloch commented on src/test/scala/mesosphere/marathon/core/task/tracker/impl/TaskOpProcessorImplTest.scala:
```diff
@@ -411,6 +413,114 @@ class TaskOpProcessorImplTest
+    f: Fixture,
+    createTaskState: (PathId) => Task,
+    createStateOp: (Task) => TaskStateOp,
+    createStateChange: (Task) => TaskStateChange): Unit = {
```
Subjective:

I think that I (personally) would have understood the test better if the parameters were called

* initialTaskState
* stateOp
* expectedStateChange

The "create" prefix would not be necessary for me. A diferrentiation between expected and given state would help.

kolloch replied:
And/Or maybe documentation for this method?

kolloch commented on src/main/scala/mesosphere/marathon/core/task/tracker/impl/TaskTrackerActor.scala:
```diff
@@ -23,15 +28,15 @@ object TaskTrackerActor {
 
-  /** Describes where and what to send after an update event has beend processed by the [[TaskTrackerActor]]. */
+  /** Describes where and what to send after an update event has been processed by the [[TaskTrackerActor]]. */
   private[impl] case class Ack(initiator: ActorRef, msg: Any = ()) {
```
Move to the side where it is used

meichstedt replied:
re-using it it in the TaskTracker again, so it's fine

kolloch commented on src/main/scala/mesosphere/marathon/core/task/tracker/impl/TaskTrackerActor.scala:
```diff
@@ -125,13 +131,18 @@ private class TaskTrackerActor(
+      case msg @ TaskTrackerActor.StateChanged(change) =>
+        val taskAfterChange = change.stateChange match {
+          case TaskStateChange.Update(task, _) => Some(task)
+          case _                               => None
```
This catch-all case is very dangerous. What about "Fail" And "Noop"?

meichstedt replied:
nice one. I've changed this to only `become` in case of Update/Expunge

kolloch commented on src/main/scala/mesosphere/marathon/core/task/tracker/impl/TaskTrackerActor.scala:
```diff
@@ -125,13 +131,18 @@ private class TaskTrackerActor(
+
+        import context.dispatcher
+        updateStepProcessor.process(change).map { _ =>
+          originalSender ! (())
```
you need to handle the failure case here. (and I'd prefer foreach for side effects)

I am not sure how to deal with errors here. I think for now, we can log errors and proceed anyway, since we use "ContinueOnError" on the steps anyway and are apparently fine with that strategy.

meichstedt replied:
foreaching now and recovering with a `log.warn`

kolloch commented on src/main/scala/mesosphere/marathon/core/task/tracker/impl/TaskTrackerUpdateStepProcessorImpl.scala:
```diff
@@ -0,0 +1,41 @@
+
+import scala.concurrent.{ ExecutionContext, Future }
+
+private[tracker] class TaskTrackerUpdateStepProcessorImpl(
```
I like this separate thanks. 

A documentation oneliner, please?

meichstedt replied:
A two-liner!

kolloch commented:
Strange, I looked at the specific commits and commented there. I hope that the comments are now not lost. Important stuff that I remember:

* TaskTrackerActor: in handling a StateChanged, the match on TaskStateChange has a default clause. That's dangerous, since it would cause a task removal for Failure/Noop.

Style issue:

* TaskTrackerActor.Ack is not used in TaskTrackerActor and can be made more specific. Then you can avoid using a redundant tuple in TaskOpProcessorImpl.

kolloch commented:
Nearly there...

meichstedt commented:
The comments are still visible, and I double-checked against the notification emails. Some comments have become irrelevant because the related (test) code has been changed by removing them (e.g. `verifyStepsAreProcessed` in `TaskOpProcessorImpl` has been removed because the steps are now processed via the Stepprocessor and I verify that that is called in `TaskTrackerActorTest`)

meichstedt commented:
Thanks @kolloch, I addressed your comments!

kolloch commented on src/main/scala/mesosphere/marathon/core/task/tracker/impl/TaskTrackerActor.scala:
```diff
@@ -131,16 +132,29 @@ private class TaskTrackerActor(
+            becomeWithUpdatedApp(task.appId)(task.taskId, newTask = None)
+
+          case _: TaskStateChange.NoChange |
+            _: TaskStateChange.Failure =>
```
:+1 for listing all choices explicitly

kolloch commented on src/main/scala/mesosphere/marathon/core/task/tracker/impl/TaskTrackerActor.scala:
```diff
@@ -29,14 +30,14 @@ object TaskTrackerActor {
+  private[impl] case class Ack(initiator: ActorRef, stateChange: TaskStateChange) {
     def sendAck(): Unit = {
-      initiator ! msg
+      initiator ! stateChange
```
I wonder if a "Failed" should be mapped to Status.Failed. Combined with an ask, the Future with then fail instead of just returning a change that includes the failure. That would be closer to what happened before.

What do you think?

meichstedt replied:
sgtm – changed it.

kolloch commented:
Test failures -- otherwise looks good.

Can you investigate? Test failures might relate to timing issues.

aquamatthias commented:
Thanks!

```




                             ________
                            `---.     `.
                                 \      `.
                                  )_______`.    
                                .'        //`---...___
                               /         ||    // ||  `-._
                            )`-|   =//=  ||   ||  ||  / ).`.
       _............_       ).-|         ||   ||  `........'`-._   (o)
    .-'         `.----`.   _...'.....__  ||   ||  _____      ||-\__.'
  .' | Inter- |  ).---.)  /_______..--'  ||   ||  -----    _ ||_/
.'_  |Plantery|  ||   ||  `-------'      ||   ||   =\\=    \_.'
|  | | Travel |  |'---'|    )`-|         ||   ||       _..-'
'--'_____________|_____|    ).-| =\\=    ||    \\  _.-' || 
|[]--------------/     /   __==\          \\   _.-'    /o'\
 \     .--.     /    _...--''   '-.__......_.-'        \__/
  `-._//'o\\___.'---''             \     .'
  LGB  \__/  -'                    /    /
                              ___.'    /
                              `-------'


```
